### PR TITLE
docs: Phase 4a — geometry-refs implicit-diff spike

### PR DESCRIPTION
**Pull request by ericchansen**
## Phase 4a of the roadmap — geometry-refs implicit-diff spike

The last remaining ❌ in the [differentiability matrix](https://github.com/ericchansen/q2mm/blob/master/docs/how-it-works/architecture.md#differentiability-status) is the geometry category (`bond_length`, `bond_angle`, `torsion_angle`), which needs the gradient of the relaxed geometry `x*(p)` with respect to the force-field parameters. Naive autodiff through the inner minimizer is expensive and brittle; the implicit function theorem gives a closed-form VJP that avoids it.

This PR is the **spike** that picks the implementation approach **before** Phase 4b writes the full production path. It does not change any runtime code — it's a runnable prototype plus a memo.

### What's in this PR

- `scripts/spike_geom_implicit_diff.py` — runnable prototype comparing two approaches on a triatomic toy system whose relaxed geometry has a closed form, so every gradient can be checked against an exact analytical derivative.
- `docs/how-it-works/geometry-refs-spike.md` — memo with three tables (accuracy vs inner tol, wall time, Hessian ill-conditioning) and the Phase 4b decision.
- `docs/how-it-works/architecture.md` — link the geometry-refs rows of the differentiability matrix to the spike.
- `properdocs.yml` — add the spike page to the nav.

### Decision

**Option A — `jaxopt.LBFGS(..., implicit_diff=True)`.**

Already a dependency. Zero custom VJP code to own. Same machine-precision accuracy as a hand-rolled `jax.custom_vjp`. Composable with the existing `jax.jit` / `jax.vmap` patterns used by `JaxOptOptimizer` and `JaxMultiStartOptimizer`.

### Summary of numbers (CPU, float64, 4-param toy)

| comparison                   | result |
| :---                         | :---   |
| A vs closed-form             | max err `9.7e-17` |
| B (custom_vjp) vs closed-form | max err `4.7e-16` |
| FD (`ε=1e-5`) vs closed-form | max err `1.8e-12` |
| A wall time                  | ~20–30 µs / gradient |
| B wall time                  | ~20–30 µs / gradient |

Inner-solver tolerance (`1e-3` → `1e-12`) does not affect outer gradient accuracy on a convex quadratic — both approaches rely only on `∇_x E(x*) ≈ 0`. Both degrade identically when `H` becomes ill-conditioned (it's a mathematical property, not an implementation choice).

### Watch-outs carried forward to Phase 4b

- Inner-solver non-convergence → needs `ok` flag + penalty fallback.
- Hessian ill-conditioning (`cond(H) > 1e4`) → Tikhonov regularization.
- `arccos` at boundary → clip cos input to `[-1+ε, 1-ε]` for collinear geometries.

### Testing

- `ruff check` + `ruff format --check` clean.
- Core test suite: 606 passed, 261 skipped, 333 deselected.
- Spike script runs end-to-end in < 1 minute on CPU.

Refs #152, #243

### CODE REVIEW

File: scripts/spike_geom_implicit_diff.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+    for tol in [1e-3, 1e-6, 1e-9, 1e-12]:
+        loss_a = jax.jit(make_loss(make_option_a(tol=tol), refs))
+        loss_b = jax.jit(make_loss(make_option_b(tol=tol), refs))
+        g_a = jax.grad(loss_a)(p)
+        g_b = jax.grad(loss_b)(p)
+        for name, g in [("A: jaxopt", g_a), ("B: custom", g_b)]:
+            err = float(jnp.max(jnp.abs(g - g_exact)))
+            rel = float(jnp.max(jnp.abs((g - g_exact) / g_exact)))
+            print(f"{tol:>10.0e} | {name:>10} | {err:>12.3e} | {rel:>12.3e}")
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply scripts/spike_geom_implicit_diff.py:215]: Fixed in 81585d4. Denominator now floored at 1e-30 (elementwise max) so a hypothetical zero in the closed-form gradient won't produce inf/nan.

File: scripts/spike_geom_implicit_diff.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    loss_ref = jax.jit(make_loss(make_option_a(tol=1e-12), refs))
+    for eps in [1e-3, 1e-5, 1e-7]:
+        g_fd = np.zeros(4)
+        for i in range(4):
+            dp = np.zeros(4)
+            dp[i] = eps
+            g_fd[i] = (loss_ref(p + dp) - loss_ref(p - dp)) / (2 * eps)
+        err = float(np.max(np.abs(g_fd - np.asarray(g_exact))))
+        rel = float(np.max(np.abs((g_fd - np.asarray(g_exact)) / np.asarray(g_exact))))
+        print(f"{eps:>10.0e} | {'FD':>10} | {err:>12.3e} | {rel:>12.3e}")
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply scripts/spike_geom_implicit_diff.py:229]: Fixed in 81585d4. Same 1e-30 floor applied to the FD rel-err denominator.

File: scripts/spike_geom_implicit_diff.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+    for k1 in [2.0, 0.1, 1e-3, 1e-5]:
+        p = jnp.array([k1, 1.5, 3.0, 1.2])
+        g_exact = closed_form_grad(p, refs)
+        x_star = closed_form_geometry(p)
+        h_mat = jax.hessian(energy, argnums=0)(x_star, p)
+        cond = float(jnp.linalg.cond(h_mat))
+        g_a = jax.grad(jax.jit(make_loss(make_option_a(tol=1e-10), refs)))(p)
+        g_b = jax.grad(jax.jit(make_loss(make_option_b(tol=1e-10), refs)))(p)
+        err_a = float(jnp.max(jnp.abs(g_a - g_exact)))
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply scripts/spike_geom_implicit_diff.py]: Fixed in 81585d4. Hoisted the jaxopt/custom grad fns and the hessian jit out of the loop; only p changes per iteration, so Table 3 now reflects steady-state runtime cost without per-iter recompilation.

File: docs/how-it-works/geometry-refs-spike.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    return run_inner_minimizer(energy, x0, p).x   # any minimizer
+
+def relax_fwd(p): x = relax(p); return x, (x, p)
+
+def relax_bwd(saved, g):
+    x, p = saved
+    H = jax.hessian(energy, argnums=0)(x, p)                 # ∂²E/∂x²
+    M = jax.jacfwd(jax.grad(energy, argnums=0), argnums=1)(x, p)  # ∂²E/∂x∂p
+    lam = jnp.linalg.solve(H.T, g)
+    return (−(lam @ M),)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply docs/how-it-works/geometry-refs-spike.md]: Fixed in 81585d4. Replaced the Unicode minus with ASCII - in the code example.
